app/TuistApp/Sources/Services/AppCredentialsService.swift

```diff
@@ -0,0 +1,81 @@
+import Foundation
+import Mockable
+import TuistServer
+import TuistSupport
+
+@Mockable
+protocol AppCredentialsServicing: ObservableObject {
+    var authenticationState: AuthenticationState { get }
+    var accountHandle: String? { get }
+
+    func loadCredentials()
+    func login() async throws
+    func logout() async throws
+    func updateAuthenticationState() async throws
+}
+
+final class AppCredentialsService: AppCredentialsServicing {
+    @Published
+    private(set) var authenticationState: AuthenticationState = .loggedOut
+
+    var accountHandle: String? {
+        switch authenticationState {
+        case let .loggedIn(accountHandle):
+            accountHandle
+        case .loggedOut:
+            nil
+        }
+    }
+
+    private let appStorage: AppStoring
+    private let serverSessionController: ServerSessionControlling
+    private let serverURLService: ServerURLServicing
+
+    init(
+        appStorage: AppStoring = AppStorage(),
+        serverSessionController: ServerSessionControlling = ServerSessionController(),
+        serverURLService: ServerURLServicing = ServerURLService()
+    ) {
+        self.appStorage = appStorage
+        self.serverSessionController = serverSessionController
+        self.serverURLService = serverURLService
+    }
+
+    func loadCredentials() {
+        authenticationState = (try? appStorage.get(AuthenticationStateKey.self)) ?? .loggedOut
+    }
+
+    func login() async throws {
+        let serverSessionController = serverSessionController
+        let serverURL = serverURLService.serverURL()
+        try await withTimeout(
+            .seconds(2 * 60),
+            onTimeout: {
+                logger.warning("Login timed out")
```

Comment:
How are these errors handled?

Reply:
On timeout, it's silently swallowed. I think it's fine as there's no running indication that there's a login in progress, so we just kill it. If the user took longer than two minutes, they can click on `Log in` again in the menu bar.